oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+  * Example: Submission blackout dates - don't allow users to submit answers during the specified time period.
+
+* ``Scope.context_preferences`` (UserScope.ONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: Enable or disable notifications about new discussion posts in this course
+  * Example: Video playback speed for this course
+  * Example: An XBlock in which students "accumulate building blocks over the course of the semester" (`PLAT-325`_)
+
+* ``Scope.context_shared_preferences`` (UserScope.ONE, BlockScope.CONTEXT_ALL)
+
+  * Example: Allow users to self-assign themselves into groups, which affects multiple types of problem XBlock (ORA, Problem Builder)?
+  * This seems to be the least-compelling new Scope, and may be dropped from the proposal if compelling use cases cannot be found.
+
+.. _PLAT-325: https://openedx.atlassian.net/browse/PLAT-325
+
+In the Open edX platform, it is expected that Studio could implement a new UI that displays all ``Scope.context_settings`` and ``Scope.context_shared_settings`` fields from any XBlocks enabled for a given course, allowing authors to configure those settings for that course. Likewise, the LMS could implement a "Preferences" tab for each course that allows students to edit all ``Scope.context_preferences``/``Scope.context_shared_preferences`` fields for XBlocks used in that course. (Though in the future, XBlocks should be able to opt-out of these auto-generated settings UIs on a per-field basis, and instead update those settings via their own UX in views like ``studio_view``, ``student_view``, or the separately proposed ``course_view``. This is expanded upon below in "Open Questions".)
+
+
+Rationale
+=========
+
+Many of the ideas in this proposal are driven by the desire for new types of functionality to be implemented via XBlocks, and thus to be pluggable and easily installed, enabled, or disabled for each Open edX installation or each course.
+
+The idea of naming the new scope "context" comes from other technologies such as LTI, where the term "context" is used analogously. For example, when embedding an LTI tool into the Open edX LMS, the LMS will pass the current course ID as the LTI "context" parameter.
+
+Previous discussions of note include:
+
+* https://openedx.atlassian.net/browse/PLAT-325
+* https://groups.google.com/d/msg/edx-code/ywjXV0wzQiw/FRzaK5nTAgAJ
+* On 2016-07-08, the author met with several stakeholders from edX (Calen Pennington, David Ormsbee, and Robert Raposa) to discuss course-scoped XBlock fields. The discussion tended toward using "context" as an additional dimension (see "Context as a New Scope Dimension" in "Rejected Alternatives"), but we did not think of any use cases for 8 of the 12 additional scopes that would be created by defining an additional dimension.
+
+Other related proposals of note include:
+
+* https://github.com/edx/XBlock/pull/317 - XBlock field sharing
```

Comment:
https://openedx.atlassian.net/wiki/display/AC/Feature+Plugins+for+edX+Platform#FeaturePluginsforedXPlatform-Plug-inArchitecture